### PR TITLE
refactor(sdk)!: Split strongly-typed state event functions into two

### PR DESCRIPTION
**Pull request by jplatte**
Another on of those refactorings I've been thinking about for a while. @Hywan just reminded me of this possibility.

I'm not too sure about the upgrade path here. It should be simple enough for people who read the changelog, but I think we have experienced many people *not* doing that..

Looking for feedback on

* Is this a good idea?
* Naming
* Upgrade path

I also posted a very small related Ruma PR: https://github.com/ruma/ruma/pull/1259

### CODE REVIEW

**Comment by Hywan** [comment]
> Is this a good idea?

I believe yes.

> Naming

Yes, I like the `_for_key` suffix.

> Upgrade path

It's fine to break the code since we are in 0.x right?

**Comment by gnunicorn** [comment]
> Upgrade path

maybe we want to start putting upgrade guides like these into a `Upgrades-0.5-to-0.6.txt`-file and add these to the PR? Like, this is straight forward enough, but the changelog generator would miss this ...

**Review by gnunicorn — APPROVED** [review]
yeah, sure, why not.

File: crates/matrix-sdk/src/room/common.rs [hunk 1]
```diff
@@ -819,22 +820,49 @@ impl Common {
...
+    /// let member_event: Option<Raw<SyncRoomMemberEvent>> = room
+    ///     .get_state_event_static_for_key(user_id!("@alice:example.org"))
+    ///     .await?;
+    /// # anyhow::Ok(())
+    /// # };
+    /// ```
+    pub async fn get_state_event_static_for_key<C, K>(
         &self,
-        state_key: &str,
+        state_key: &K,
```
Comment (poljar) [inline crates/matrix-sdk/src/room/common.rs:857]: What's the rationale for making this generic instead of keeping `&str`?

Comment (jplatte) [reply crates/matrix-sdk/src/room/common.rs:857]: This way you can pass `&UserId` for a member event, `&RoomId` for a space parent event and such, without `.as_str()`. I actually found a "bug" through this where a test was passing an empty string for getting a member event, which would never return an actual event in the real world.

**Review by poljar — APPROVED** [review]

**Comment by jplatte** [comment]
Somebody should probably review the documentation changes :)

**Review by gnunicorn — APPROVED** [review]
docs look fine to me

**Pull request merged**